### ISSUE TITLE
reportUnusedFunction triggers for inner functions which are captured by decorators

### ISSUE THREAD

**Issue by mezuzza** [microsoft/pyright#7771]
**Describe the bug**
When using a decorator to capture functions in a container which is returned, functions are incorrectly reported as being unused.

**Code or Screenshots**
```python
from typing import Callable


class Registry:
    def __init__(self) -> None:
        # Any container of callables works here, this is just one illustration
        self.registry: list[Callable[[], int]] = []

    def register(self, f: Callable[[], int]) -> Callable[[], int]:
        self.registry.append(f)
        return f

def setup_registry():
    registry = Registry()

    @registry.register
    def foo(): # pyright: foo is not accessed
        return 3

    @registry.register
    def bar(): # pyright: bar is not accessed
        return 4

    return registry
    
r = setup_registry()
for f in r.registry:
  print(f())
```

Note that foo and bar are accessible from outside through the registry. I know this is a strange case, but you run into this in real world scenarios when doing dependency injection into ASGI/WSGI applications (fastapi/flask) e.g.

```python
def setup_route(some_dependency: dict[str, str]):
    router = APIRouter()

    @router.get("/")
    def health_check():
        return some_dependency["version"]

    return router
```

Two simple workarounds:
1. Use the method instead of the decorator: `router.get("/")(health_check)`.
2. `del health_check` afterwards which is what I currently do.

I don't know a good solution here. Here are some options:
1. Require functions to be prefixed with `_` to trigger reportUnusedFunction
    1. This is likely annoying since most people don't write their inner functions with a leading `_`.
2. Don't trigger on functions with decorators
    1. Some valid decorators could cause this to be underreported e.g. `@lru_cache`.
3. Allow ignoring the issue with `# type: ignore [reportUnusedFunction]`
    1. This doesn't currently work, but it could alleviate the problem.
    2. It's still noisy and unintuitive.
4. No change
    1. There are workarounds
    2. Feels odd that manually calling a decorator fixes the problem and kind of makes a common python pattern annoying to use.

I recognize that this isn't a super important problem, but thought I'd have it captured.

**VS Code extension or command-line**
python: 3.12.2
pyright: 1.1.356

**Comment by erictraut** [comment]
Pyright is working as intended here. When you apply a decorator to a function, you're effectively assigning the return result of the decorator back to a local symbol, and this symbol is not referenced.

```python
@router.get("/")
def health_check(): ...
```

is roughly equivalent to:

```python
def _health_check(): ...

health_check = router.get("/")(_health_check)
```

The local symbol `health_check` is not being used.

If your intent is to simply register `health_check`, then you could do the following instead:

```python
def health_check(): ...
router.get("/")(_health_check)
```

I'll note that pyright doesn't emit any diagnostics for this condition unless you've enabled `reportUnusedFunction`. If you don't want these diagnostics, you can choose not to enable this diagnostic rule or suppress it on a case-by-case basis.

Pyright does emit a "tagged hint" in this case, which tells the client that it should display the identifier in a "grayed-out" manner. Some clients don't properly display tagged hints and instead display them like normal diagnostics. If you're using such a client, you may want to set the language server setting "pyright.disableTaggedHints" to false.

**Comment by samuelcolvin** [comment]
This is very unfortunate.

FastAPI is very popular, and the decorator pattern shown here is even more widely used.

This behaviour seems both opinionated and unhelpful in many common scenarios, @erictraut would you consider reversing this position? Or at least provide a way to opt out of this behaviour without muting all `reportUnusedFunction` errors.

**Comment by erictraut** [comment]
@samuelcolvin, are you suggesting that any decorator applied to any function or method always mark that function or method as implicitly used? I don't think that would be correct behavior.

The decorator here is being used in an unusual manner. The vast majority of decorators don't perform dynamic registrations and should not imply usage of the decorated function.

I'll also note that pyright's `reportUnusedFunction` check is off by default. It is enabled only in "strict" mode or if you manually enable it. Strict-mode checks are often more opinionated than standard-mode checks.

**Comment by hmc-cs-mdrissi** [comment]
My experience is decorators very commonly do registration of some form. Fastapi’s practice is not original to them. Flask/similar webframeworks also rely on decorators for registration of routes. That already probably leads to significant percent of users who work on web services in python where decorators do registration.

My config library also relies on decorators for registration although often the registered functions/classes are never called directly and only through registry so I just name them with underscore prefix otherwise I’d probably get a fair amount of false positives in my codebase as well.

**Comment by mezuzza** [comment]
+1 on experiencing decorators very commonly doing some form of registration.

I would also like to say that strict mode and reporting unused functions in general is a very useful behavior too opt in to. Yes it's true that we could disable all of these reports, but that's really throwing the baby out with the bath water.

**Comment by samuelcolvin** [comment]
To be fair, this doesn't apply in the module namespace, where FastAPI "register" decorated functions are normally defined.

I therefore think it's probably fine as-is.

**Comment by samuelcolvin** [comment]
For anyone else coming  to this and having particular problems in tests, you can add the following to `pyproject.toml`

```toml
[tool.pyright]
typeCheckingMode = "strict"
...
# see https://github.com/microsoft/pyright/issues/7771, we don't want to error on decorated functions in tests
# which are not otherwise used
executionEnvironments = [
  { root = "tests", reportUnusedFunction = false },
]
```

**Comment by YanSte** [comment]
Thanks @samuelcolvin

**Comment by hirak99** [comment]
Setting `reportUnusedFunction = False` works, but it covers a wider area than we need, i.e. now it will stop checking for unused methods on every file.

I recommend instead add only to the file where you need it.

```python
# pyright: reportUnusedFunction=false

@app.route("/")
def serve_index():
  ...
```

It would be better however if there was a new directive that narrows the scope further to decorators, "reportUnusedDecoratedFunction".

**Comment by dmontagu** [comment]
I agree that it would be nice if this was a separate directive, but another possible improvement would be to not emit this error if the decorator is annotated as returning `None` rather than the (callable) input type, as this more or less clearly indicates the output of the decorator is not intended to be used. (Or better yet, providing some way for the API author to indicate that the decorator "consumes" the decorated function while still returning the value; but I'd settle for just suppressing when it returns `None`.)

This wouldn't help reduce the nuisance for decorators that really do need to return the decorated function (such as FastAPI's `@app.get` etc.), but it would provide a way to define an API that uses decorators to "consume" a defined function without emitting type errors.

Yes, users could always do the `def f(): ...; my_decorator(f)` thing, but imo there are frequently readability benefits to a decorator even in this "consuming the function" scenario — when there is a decorator at the top, you see an indication right away about what the function is going to be used for _before_ you start skimming the definition; if you call the decorator on the function afterward, you have to reach the bottom of the function definition before getting that sense of how it's being used.

**Comment by mezuzza** [comment]
That suggestion has a really nice benefit that one can simply wrap non-compliant decorators manually allowing folks to opt into this behavior when necessary
